framework/SupervisedLearning/ARMA.py

```diff
@@ -1102,6 +1152,120 @@ def finalizeGlobalRomSegmentEvaluation(self, settings, evaluation, weights=None)
+    for i in range(windowType):
+      windowRange={}
+      bg_P=(windows[i]['window'][0]-1)%period
+      end_P=(windows[i]['window'][1]+2)%period
+      timeInd=np.arange(len(self.pivotParameterValues))
+      bg_P_ind  = np.where(timeInd%period==bg_P )[0].tolist()
+      end_P_ind = np.where(timeInd%period==end_P)[0].tolist()
+      if bg_P_ind[0]>end_P_ind[0]:
+        tail=end_P_ind[0]
+        end_P_ind.pop(0)
+        end_P_ind.append(tail)
+      windowRange['bg']=bg_P_ind
+      windowRange['end']=end_P_ind
+      rangeWindow.append(windowRange)
+    return rangeWindow
+
+  def _peakGroupWindow(self,signal,windowDict):
+    """
+      Collect the peak information in to groups, define the residual signal
+      @ In, signal, np.array(float), signal to transform
+      @ In, windowDict, dict, dictionary for specefic target peaks
+      @ Out, groupWin, list, list of dictionaries which store the peak information
+      @ Out, maskRes, np.array, boolean mask where is the residual signal
+    """
+    groupWin = []
+    maskRes = signal == signal
+    rangeWindow = self.rangeWindow(windowDict)
+    low = windowDict['threshold']
+    windows = windowDict['windows']
+    for i in range(len(windowDict['windows'])):
+      bg  = rangeWindow[i]['bg']
+      end = rangeWindow[i]['end']
+      peakInfo   = {}
+      indLocal   = []
+      ampLocal   = []
+      for j in range(min(len(bg), len(end))):
+        ##FIXME this might ignore one window
+        bg_local = bg[j]
+        end_local = end[j]
+        peak, height = self._peakPicker(signal[bg_local:end_local], low=low)
+        if len(peak) ==1:
+          indLocal.append(int(peak))
+          ampLocal.append(float(height))
+          mask_bg=int(peak)+bg_local-int(np.floor(windows[i]['width']/2))
+          mask_end=int(peak)+bg_local+int(np.ceil(windows[i]['width']/2))
+          maskRes[mask_bg:mask_end]=False
+        elif len(peak) >1:
+          indLocal.append(int(peak[np.argmax(height)]))
+          ampLocal.append(float(height[np.argmax(height)]))
+          mask_bg=int(peak[np.argmax(height)])+bg_local-int(np.floor(windows[i]['width']/2))
+          mask_end=int(peak[np.argmax(height)])+bg_local+int(np.ceil(windows[i]['width']/2))
+          maskRes[mask_bg:mask_end]=False
+      peakInfo['Ind'] = indLocal
+      peakInfo['Amp'] = ampLocal
+      groupWin.append(peakInfo)
+    return groupWin , maskRes
+
+  def _transformBackPeaks(self,signal,windowDict):
+    """
+      Transforms a signal by regenerate the peaks signal
+      @ In, signal, np.array(float), signal to transform
+      @ In, windowDict, dict, dictionary for specefic target peaks
+      @ Out, signal, np.array(float), new signal after transformation
+    """
+    groupWin = windowDict['groupWin']
+    windows  = windowDict['windows']
+    rangeWindow=self.rangeWindow(windowDict)
+    for i in range(len(windowDict['windows'])):
+      prbExist = len(groupWin[i]['Ind'])/len(rangeWindow[i]['bg'])
```

Comment:
Some comments on what you're doing here (check if we should add a peak, determine peak amplitude, etc) would be helpful for readability.